### PR TITLE
feat: add LeRobot policy wrapper with save_pretrained, checkpoint converter, and equivalence tests

### PR DESCRIPTION
**Pull request by samet-akcay**
## Description

Thin Lightning wrappers around LeRobot policies. All heavy lifting (optimizer, loss, normalization) stays in LeRobot — physicalai only handles the Lightning ↔ LeRobot boundary.

This PR adopts a **two-tier policy contract** (per reviewer feedback):

- **8 first-class named wrappers** with explicit support: `act`, `diffusion`, `groot`, `pi0`, `pi05`, `pi0_fast`, `smolvla`, `xvla`.
- **Universal escape hatch** via `LeRobotPolicy(policy_name=...)` accepting any LeRobot policy name. Unsupported names emit a one-time `UserWarning` — no warranty that they will work.

The named subset is deliberately curated. Equivalence is **only** guaranteed for the validated subset (see Validation below); named-but-unvalidated wrappers (`groot`, `xvla`) carry documented limitations and the universal escape hatch is best-effort.

---

### 1. Instantiate any LeRobot policy

```python
from physicalai.policies.lerobot import LeRobotPolicy, ACT, PI05

# Universal wrapper — any LeRobot policy by name (named ones get the equivalence
# guarantee for the validated subset; others are best-effort + UserWarning)
policy = LeRobotPolicy(policy_name="pi05", optimizer_lr=1e-4)

# Named wrapper — full type safety and IDE autocomplete
policy = ACT(dim_model=512, chunk_size=100, n_action_steps=100)

# Convenience alias
policy = PI05(optimizer_lr=1e-4)
```

### 2. Create from a LeRobot dataset

Automatically infers input/output features from dataset metadata:

```python
from lerobot.datasets.lerobot_dataset import LeRobotDataset
from physicalai.policies.lerobot import LeRobotPolicy

dataset = LeRobotDataset("lerobot/aloha_sim_transfer_cube_human")
policy = LeRobotPolicy.from_dataset("pi05", dataset)
# Or pass a repo ID directly (only metadata is fetched):
policy = LeRobotPolicy.from_dataset("pi05", "lerobot/aloha_sim_transfer_cube_human")
```

### 3. Load pretrained from HuggingFace Hub

```python
from physicalai.policies.lerobot import ACT

policy = ACT.from_pretrained("lerobot/act_aloha_sim_transfer_cube_human")
```

### 4. Train with Lightning

```python
from physicalai.data import LeRobotDataModule
from physicalai.train import Trainer

datamodule = LeRobotDataModule(repo_id="lerobot/aloha_sim_transfer_cube_human")
trainer = Trainer(max_steps=50_000, accelerator="gpu", precision="bf16-mixed")
trainer.fit(policy, datamodule=datamodule)
```

### 5. Export & share

```python
# Save locally (HuggingFace-compatible: config.json + model.safetensors)
policy.save_pretrained("./pi05_trained")

# Push to Hub
policy.save_pretrained("./pi05_trained", repo_id="username/pi05-aloha")

# Round-trip back to native LeRobot
from lerobot.policies.pretrained import PreTrainedPolicy
native = PreTrainedPolicy.from_pretrained("./pi05_trained")
```

### 6. Checkpoint conversion

Bidirectional conversion between Lightning `.ckpt` and LeRobot safetensors:

```python
from physicalai.policies.lerobot import lightning_to_lerobot, lerobot_to_lightning

# Lightning → LeRobot
lightning_to_lerobot("checkpoints/epoch=10.ckpt", "output/lerobot_model/")

# LeRobot → Lightning
lerobot_to_lightning("output/lerobot_model/", "output/restored.ckpt")
```

---

## Validation

Two registries describe wrapper coverage:

- **`SUPPORTED_POLICIES`** (8) — named first-class wrappers exposed as `NamedLeRobotPolicy` subclasses.
- **`VALIDATED_EQUIVALENCE_POLICIES`** (6) — subset with measured wrapper-vs-native numerical equivalence.

Anything outside `SUPPORTED_POLICIES` is reachable only via `LeRobotPolicy(policy_name=...)` and emits a one-time `UserWarning`.

Equivalence is validated across multiple tiers (`library/tests/unit/policies/test_lerobot.py` + `library/tests/integration/test_lerobot_wrapper_equivalence.py`):

| Tier | Class | What | Tolerance | Where |
|---|---|---|---|---|
| Unit (CPU, fp32) | `TestLeRobotPolicyEquivalence` | Forward / gradient / 5-step weights | `rtol=atol=1e-6` | unit |
| 1 — Fast-dev-run | `TestFastDevRunEquivalence` | Single-step loss vs native | `rtol=1e-5` | integration |
| 2 — Multi-step | `TestMultiStepTrainerEquivalence` | 10-step loss trajectories | `rtol=1e-5` | integration |
| 3 — Gradient | `TestGradientEquivalence` | First-step parameter gradients | `rtol=1e-5` | integration |
| 4 — Weights | `TestWeightEquivalenceAfterTraining` | Post-training (10 step) weights | `rtol=5e-5` (calibrated) | integration |
| 5 — Regression | `TestRegressionTraining` | 50-step loss decreases + trajectory | `rtol=1e-3` | integration (`@slow`) |

The `5e-5` weight tolerance is calibrated from observed 1.58e-5 max drift on 8/2.36M elements of `model.backbone.layer4.0.conv2.weight` after 10 ACT steps; ~3x buffer keeps it tight enough to catch optimizer-state divergence.

The integration tier auto-detects accelerators via `physicalai.devices.get_device()` (CUDA > XPU > CPU), so Intel GPUs run the same suite.

#### Per-policy validation status

| Policy | Unit (CPU) | Integration (CUDA/XPU) | Notes |
|---|---|---|---|
| `act`, `diffusion`, `smolvla` | ✅ | ✅ | Validated CPU + accelerator tiers |
| `pi0`, `pi05`, `pi0_fast` | — | ✅ | VLA, accelerator + bf16-mixed only |
| `groot` | — | xfail | Upstream hardcodes `flash_attention_2` |
| `xvla` | — | xfail | Requires explicit `vision_config` kwarg |

Tracked limitations register as pytest `xfail(reason=...)` rather than silent skips, so any future fix surfaces as `XPASS` and prompts moving the policy into `VALIDATED_EQUIVALENCE_POLICIES`. An invariant test (`test_named_wrappers_share_policy_name_with_base`) asserts `VALIDATED_EQUIVALENCE_POLICIES` and `_EQUIVALENCE_XFAIL_REASONS` are disjoint.

### Unit tests

`library/tests/unit/policies/test_lerobot.py` — full unit suite passes (967 passed, 4 skipped on this branch). The 4 skips cover optional extras and CPU-incompatible VLA configurations.

---

## Wrapper vs Native LeRobot

The wrapper produces equivalent training dynamics to native LeRobot under the tolerances above. The actual comparison harness (see `_make_wrapper_and_native` in `library/tests/integration/test_lerobot_wrapper_equivalence.py`) builds the native policy by deep-copying the wrapper's inner policy and reusing the wrapper's optimizer preset and preprocessor — that is what makes "same weights, same batches, same optimizer" actually true.

### With physicalai wrapper

```python
from physicalai.data import LeRobotDataModule
from physicalai.policies.lerobot import LeRobotPolicy
from physicalai.train import Trainer

datamodule = LeRobotDataModule(repo_id="lerobot/aloha_sim_transfer_cube_human")
datamodule.setup("fit")

wrapper = LeRobotPolicy.from_dataset("pi05", datamodule.train_dataset)

trainer = Trainer(max_steps=50, accelerator="gpu", precision="bf16-mixed")
trainer.fit(wrapper, datamodule=datamodule)
```

### With native LeRobot (same model, same weights, same optimizer, same preprocessing)

```python
import copy, torch
from physicalai.devices import get_device

device = get_device()  # auto: cuda > xpu > cpu

# Same starting weights as the wrapper
native = copy.deepcopy(wrapper.lerobot_policy).to(device).train()

# Same optimizer preset the wrapper would build internally
native_optimizer = wrapper._config.get_optimizer_preset().build(native.get_optim_params())

# Same preprocessing pipeline the wrapper applies before forward()
preprocessor = wrapper._preprocessor

for step_idx, batch in enumerate(captured_batches):
    native_optimizer.zero_grad()
    torch.manual_seed(base_seed + step_idx)  # match wrapper's per-step seed
    preprocessed = preprocessor(batch)
    loss, _ = native.forward(preprocessed)
    loss.backward()
    native_optimizer.step()
```

`captured_batches` are the exact batches the Lightning Trainer fed the wrapper, recorded via a callback (`LossBatchCaptureCallback`). The preprocessor is shared because normalization stats and image transforms are stateful — feeding raw dataloader batches to native would diverge immediately.

### CODE REVIEW

File: library/src/physicalai/policies/lerobot/__init__.py [hunk 1]
```diff
@@ -167,19 +181,20 @@ def __init__(self, **kwargs) -> None:  # noqa: ANN003
...
-        super().__init__(policy_name="pi0fast", **kwargs)
+        super().__init__(policy_name="pi0_fast", **kwargs)
 
 
 __all__ = [
     "ACT",
     "PI0",
     "PI05",
     "SAC",
+    "SUPPORTED_POLICIES",
```
Comment (sovrasov) [inline library/src/physicalai/policies/lerobot/__init__.py:150]: ACT, Groot, and diffusion still have dedicated wrappers in `library/src/physicalai/policies/lerobot/`. Does it still make sense? Other policies are now wrapped via `LeRobotPolicy`

**Review by sovrasov — DISMISSED** [review]

**Review by sovrasov — APPROVED** [review]
